### PR TITLE
Update Cassandra driver to 4.x

### PR DESCRIPTION
**Pull request by emerkle826**
This PR updates the Cassandra driver to the newer 4.x version. It should address https://github.com/testcontainers/testcontainers-java/issues/2768.

### CODE REVIEW

**Comment by rnorth** [comment]
I'm a bit uncomfortable with this causing a breaking change in our APIs and transitive dependencies, but I'm not sure I can see any other way around it (short of adding an entire new class that uses the Cassandra 4.x drivers).

Will post a suggestion as a comment, though.

File: modules/cassandra/src/main/java/org/testcontainers/containers/CassandraContainer.java
Comment (rnorth) [inline modules/cassandra/src/main/java/org/testcontainers/containers/CassandraContainer.java:162]: Returning a third-party class from our convenience method is what's now given us this breaking change headache. It's a pattern we're trying to avoid with new modules for this exact reason.

Could we consider eliminating `com.datastax.*` types from the API of this class?

e.g.:
* delete the old `getCluster()` method as you've done
* do not add a `getSession` method
* add a new `getContactPoint()` method that simply returns an `InetSocketAddress`, and a `getDataCenter()` method that returns the `DATACENTER` constant

It would be slightly more work for users, e.g. they would have to do:
```
final CqlSessionBuilder builder = CqlSession.builder()
            .addContactPoint(myCassandraContainer.getContactPoint())
            .withLocalDatacenter(myCassandraContainer.getDatacenter());
        return builder.build();
```

but our public API from this class would just consist of JDK types, and we'd never have to make breaking changes again should (for example) the cassandra driver change classes/package names in v5.

If we're going to have to have a breaking change here, why not do it in a way that avoids future breaking changes?

BTW you'll notice I'm suggesting this be instance methods, not static methods - I am really quite surprised that these were ever static methods.

Comment (emerkle826) [reply modules/cassandra/src/main/java/org/testcontainers/containers/CassandraContainer.java:162]: @rnorth Thank you for the review and the comments!
I see your point about Testcontainers concerns and I think you are right. I'm a bit tied up with a few other things at the moment, but I will revisit this as soon as I can and try to implement the changes you suggest. I'll also try to update the Cassandra Testcontainer documentation if necessary.

**Comment by emerkle826** [comment]
> I'm a bit uncomfortable with this causing a breaking change in our APIs and transitive dependencies, but I'm not sure I can see any other way around it (short of adding an entire new class that uses the Cassandra 4.x drivers).
> 
> Will post a suggestion as a comment, though.

@rnorth I've taken a second stab at this. Instead of removing the existing Driver 3.x API, I've chosen to deprecate the methods that return its `Cluster` object, while adding the new methods you suggested. This prevents introducing a breaking change for now, and provides a tiny convenience for using the container with Driver 4.x. I also updated the docs to demonstrate the _new_ way of using it with the different driver versions.

However, I have left [CassandraDatabaseDelegate](https://github.com/testcontainers/testcontainers-java/blob/master/modules/cassandra/src/main/java/org/testcontainers/containers/delegate/CassandraDatabaseDelegate.java) and [CassandraQueryWaitStrategy](https://github.com/testcontainers/testcontainers-java/blob/master/modules/cassandra/src/main/java/org/testcontainers/containers/wait/CassandraQueryWaitStrategy.java) alone for now as I'm not sure the best solution to make either of those Driver agnostic. A solution could be to deprecate/remove the delegate and wait strategy, and maybe add more documentation of how to do it with each of the drivers. Or, I could duplicate everything you have currently and provide Driver 4 versions of the same classes. But I believe that is not really in the interests of Testcontainers, and probably creates more confusion that it does good.

File: modules/cassandra/src/test/java/org/testcontainers/containers/CassandraContainerTest.java
Comment (tomekl007) [inline modules/cassandra/src/test/java/org/testcontainers/containers/CassandraContainerTest.java:125]: Maybe instead of `@Test(expected = IllegalStateException.class)`, we could do a more detailed verification by catching an exception and validating that the exception message contains `contact-points` string? Or use AssertJ `assertThatThrownBy`

File: modules/cassandra/src/main/java/org/testcontainers/containers/CassandraContainer.java
Comment (tomekl007) [inline modules/cassandra/src/main/java/org/testcontainers/containers/CassandraContainer.java:228]: nice that you added it, it was missing before.

File: modules/cassandra/src/test/java/org/testcontainers/containers/CassandraContainerTest.java
Comment (emerkle826) [reply modules/cassandra/src/test/java/org/testcontainers/containers/CassandraContainerTest.java:128]: I've changed the test to verify part of the exception message here: https://github.com/testcontainers/testcontainers-java/pull/2830/files#diff-f060cde9ce304b8f1959fc12ce03a851R131-R132

**Comment by michaelyaakoby** [comment]
I'm using the changes from this pull-request for many months already so we can test using latest cassandra driver.
Unless testcontainers already added this support, it would be strange if this PR is closed.

Maybe I or others can help with this?

**Comment by pluttrell** [comment]
@rnorth Soft ping. Wondering what it would take to get this to the finish line? I'd love to start using it.

File: modules/cassandra/src/main/java/org/testcontainers/containers/CassandraContainer.java
Comment (bsideup) [inline modules/cassandra/src/main/java/org/testcontainers/containers/CassandraContainer.java:41]: this variable seems to be write-only. It does not end up in container's configuration (e.g. environment variable). Let's either make it affect the container or remove.

File: modules/cassandra/src/main/java/org/testcontainers/containers/CassandraContainer.java
Comment (bsideup) [inline modules/cassandra/src/main/java/org/testcontainers/containers/CassandraContainer.java:33]:
```suggestion
    private static final String DEFAULT_LOCAL_DATACENTER = "datacenter1";
```

File: modules/cassandra/src/main/java/org/testcontainers/containers/CassandraContainer.java
Comment (bsideup) [inline modules/cassandra/src/main/java/org/testcontainers/containers/CassandraContainer.java:181]: Why?

```suggestion
     * Get configured Cluster
```

File: modules/cassandra/src/main/java/org/testcontainers/containers/delegate/CassandraDatabaseDelegate.java
Comment (bsideup) [inline modules/cassandra/src/main/java/org/testcontainers/containers/delegate/CassandraDatabaseDelegate.java:28]: let's revert this non-relevant change to reduce the changeset

File: modules/cassandra/src/test/java/org/testcontainers/containers/CassandraContainerTest.java
Comment (bsideup) [inline modules/cassandra/src/test/java/org/testcontainers/containers/CassandraContainerTest.java:30]: typo, `BAISC_QUERY` -> `BASIC_QUERY`

File: modules/cassandra/src/test/java/org/testcontainers/containers/CassandraContainerTest.java
Comment (bsideup) [inline modules/cassandra/src/test/java/org/testcontainers/containers/CassandraContainerTest.java:123]: this seems to be testing the driver, not the container. Let's remove it

File: modules/cassandra/src/main/java/org/testcontainers/containers/CassandraContainer.java
Comment (adutra) [reply modules/cassandra/src/main/java/org/testcontainers/containers/CassandraContainer.java:181]: There is no more `Cluster` vs `Session` distinction in driver4. Your only entry point is the session, now called `CqlSession`. The word "cluster" becomes meaningless.

Comment (bsideup) [reply modules/cassandra/src/main/java/org/testcontainers/containers/CassandraContainer.java:181]: this only applies to v4, while in v3 (the currently used types that we're deprecating) is remains `Cluster`.

Comment (adutra) [reply modules/cassandra/src/main/java/org/testcontainers/containers/CassandraContainer.java:181]: Whoops you are right, my bad.

Comment (emerkle826) [reply modules/cassandra/src/main/java/org/testcontainers/containers/CassandraContainer.java:181]: Yes, one of my first commits for this changed `getCluster()` to `getSession()`, but that was reverted. I just forgot to revert the comment change.

**Comment by emerkle826** [comment]
@bsideup I've made the changes you suggested, however I'm still not sure what to do about the wait strategy (mentioned [here](https://github.com/testcontainers/testcontainers-java/pull/2830#issuecomment-667355913)). I'm not sure if that is an issue that needs to be addressed at this time.

**Comment by akhaku** [comment]
Gentle nudge for the reviewers from another user patiently waiting for the upgrade to Datastax 4 :)

**Comment by akhaku** [comment]
I think this is waiting for changes due to a review from @rnorth, but I'm not sure if his comments are applicable anymore.

**Comment by findinpath** [comment]
@emerkle826 I happen to work with cassadra java driver 4 and actually I am not quite sure why `CassandraContainer#getCluster()` convenience method is actually provided in the testcontainers-cassandra module.

I currently use testcontainers 1.15.3 and do something like this:

```
try (var session = CqlSession.builder()
        .addContactPoint(new InetSocketAddress(
                cassandraContainer.getContainerIpAddress(),
                cassandraContainer.getFirstMappedPort()))
        .withLocalDatacenter("datacenter1")
        .withKeyspace(CqlIdentifier.fromCql(TEST_KEYSPACE))
        .build()) {
 
   //use here the session for interracting with Cassandra
}
```

Note that the CqlSession is autocloseable.

I was under the impression that I have to make a PR like yours, but then I've looked over [testcontainers-elasticsearch](https://github.com/testcontainers/testcontainers-java/blob/master/modules/elasticsearch/src/main/java/org/testcontainers/elasticsearch/ElasticsearchContainer.java) and didn't see any convenience method for offering  a [RestHighLevelClient](https://www.elastic.co/guide/en/elasticsearch/client/java-rest/master/java-rest-high-getting-started-initialization.html).

It seems to me that this shouldn't be the responsibility of the container class.

**Comment by akhaku** [comment]
Just a data point: we ended up forking `org.testcontainers:cassandra` internally and merging in that change and making a few more changes to remove usage of the 3.x driver entirely, including removing the `CassandraContainer#getCluster()` method you called out.

**Comment by stefanos-kalantzis** [comment]
Any chance for this being merged? Is there some help needed? I could help with the code, but I don't have approval rights.

@rnorth ?

**Comment by MarcusLinkeHermes** [comment]
@rnorth Any updates?

**Comment by nikeee** [comment]
There is also #3952.

Any chance to get this merged? The current version clashes with the version that  Spring-Data-Cassandra uses.

Any chance to get this merged?

File: examples/cassandra-container/build.gradle
Comment (kiview) [inline examples/cassandra-container/build.gradle:18]:
```suggestion
```
Not needed, or?

**Review by bsideup — APPROVED** [review]

**Review by kiview — APPROVED** [review]

**Comment by kiview** [comment]
Merged, thank you @emerkle826 and everyone collaborating on this 🙂

**Comment by nikeee** [comment]
Thank you!

**Pull request merged**
